### PR TITLE
Feat/initial distrib constructor

### PR DESCRIPTION
**Pull request by 0xEniotna**
Issue: #60 

This is a first draft. 

In the proposed implementation, the dev would pass an array `initial_holders` of tuples `(initial_holder_address,  amount)`. 
The `initial_recipient` should be in the position 0 of the array. As `initial_recipient` is an `initial_holder`, it makes sense to add him in the array. But it implies that the sum of all amounts of the `initial_holders` array should remain equal to `initial_supply`. This is verified by line 116 `assert(initial_minted_supply <= initial_supply, 'Unruggable: max supply reached');`.
I'm also considering that the `initial_recipient` account should deposit ALL of his tokens when `launch_memecoin` is called.. Otherwise it is not possible to perform the "team allocation" check since the balance of `initial_recipient` would be greater. If `initial_recipient` does not deposit all of his token in `launch_memecoin` then I'm not performing the "max team allocation" check on him which bypasses the rule (does that make sense ?). 

Didn't write the tests yet, I wanted to have feedbacks on my implementation.

### CODE REVIEW

**Comment by 0xEniotna** [comment]
I still need to write specific tests to try tricky scenarios

File: contracts/src/tokens/memecoin.cairo [hunk 1]
```diff
@@ -64,23 +64,56 @@ mod UnruggableMemecoin {
...
     #[constructor]
     fn constructor(
         ref self: ContractState,
         owner: ContractAddress,
         initial_recipient: ContractAddress,
         name: felt252,
         symbol: felt252,
-        initial_supply: u256
+        initial_supply: u256,
+        initial_holders: Array<ContractAddress>,
```
Comment (enitrat) [inline contracts/src/tokens/memecoin.cairo:77]:
```suggestion
        initial_holders: Span<ContractAddress>,
```

always use Spans if the arrays are not modified

File: contracts/src/tokens/memecoin.cairo [hunk 2]
```diff
@@ -64,23 +64,56 @@ mod UnruggableMemecoin {
...
+        initial_holders_amounts: Array<u256>,
```
Comment (enitrat) [inline contracts/src/tokens/memecoin.cairo:78]: why note initial_holders: Array<(ContractAddress, u256)>,

**Comment by 0xEniotna** [comment]
Added tests, should be ready now

File: contracts/tests/test_unruggable_memecoin.cairo [hunk 3]
```diff
@@ -60,32 +82,61 @@ mod erc20_metadata {
         let name = 'UnruggableMemecoin';
         let symbol = 'UM';
         let initial_supply = 1000.into();
-        let contract_address = deploy_contract(owner, recipient, name, symbol, initial_supply);
+        let initial_holder_1 = contract_address_const::<44>();
+        let initial_holder_2 = contract_address_const::<45>();
+        let initial_holders = array![recipient, initial_holder_1, initial_holder_2].span();
+        let initial_holders_amounts = array![900.into(), 50.into(), 50.into()].span();
+        let contract_address =
+            match deploy_contract(
```
Comment (abdelhamidbakhta) [inline contracts/tests/test_unruggable_memecoin.cairo:90]: could we have a test utility function to make it smoother and avoid duplication ?

**Review by abdelhamidbakhta — COMMENTED** [review]
left some comments

**Review by abdelhamidbakhta — APPROVED** [review]
lgtm

**Pull request merged**